state/protocol/protocol_state/kvstore_storage.go

```diff
@@ -14,19 +14,17 @@ import (
 // supported by the current software version. There might be serialized snapshots with legacy versions
 // in the database that are not supported anymore by this software version.
 type ProtocolKVStore interface {
-	// StoreTx returns an anonymous function (intended to be executed as part of a database transaction),
-	// which persists the given KV-store snapshot as part of a DB tx. Per convention, all implementations
+	// BatchStore writes the given KV-store snapshot to the input write batch. Per convention, all implementations
 	// of `protocol.KVStoreReader` should be able to successfully encode their state into a data blob.
 	// If the encoding fails, the anonymous function returns an error upon call.
 	//
 	// Expected errors of the returned anonymous function:
 	//   - storage.ErrAlreadyExists if a KV-store snapshot with the given id is already stored.
-	StoreTx(stateID flow.Identifier, kvStore protocol.KVStoreReader) func(*transaction.Tx) error
+	BatchStore(rw storage.ReaderBatchWriter, stateID flow.Identifier, kvStore protocol.KVStoreReader) error
```

Comment:
also here, the intention was to precompute a bunch of stuff and then commit the changes in the end.

Reply:
Thanks @jordanschalm for the summary in slack and your detailed response here in the PR.

Now that I better understand, I agree with Jordan's comment that reverting back to using functors is _not_ a good idea. While I looked into Pebble's write batch a while back, I unfortunately forgot that `pebble.Batch` (hence, also `storage.ReaderBatchWriter`) are effectively deferred database writes, which are only performed once [Batch.Commit](https://github.com/onflow/flow-go/blob/ea69a7a3fefcb2806106f924476ece1cfefcfef8/storage/operation/pebbleimpl/writer.go#L108) is called.

I will go over my comments and resolve/delete those comments where I suggested to bring the functors back in the storage API.